Implement Subscription Backport Library

blangley28 opened the pull request:
- Extract the Subscription Backport work from Joe's HAPI server into its own library
- Refactor to use HAPI FHIR types instead of JSON
- Created a basic model for SubscriptionTopics
- TopicListInterceptor updated to use SubscriptionTopics instead of static file
- SubscriptionInterceptor updated to receive the triggering resource instead of waiting 15s then querying for lastUpdated
- SubscriptionInterceptor takes in ISearchClient which can be implemented for different auth schemes etc.
- SubscriptionInterceptor implements a generic topic and subscription filtering engine to determine which subscriptions to send notifications to

To test run `./gradlew publishToMavenLocal` and then run the HAPI server.
Marking as DRAFT because there are still quite a few TODOs which I plan to take a look at

Commit:
Initial library code

Commit:
Initial refactor

Commit:
Fix checkstyles

Commit:
Fix Javadoc for maven

Commit:
Topic list working

Commit:
Implement generic engine

Commit:
Check criteria before sending notification

Commit:
Fix criteria validation and UPDATE request type match

Commit:
Revert UPDATE and refactor SubscriptionHelper

Commit:
debugging

Commit:
Add CI test

Commit:
Intercept requested subscriptions and set status to active

Commit:
Set status to error if notification fails to send

Commit:
Add more checking for valid subscriptions

Commit:
Add status parameter

Commit:
Fix parameters resource

dehall commented on build.gradle:
```diff
@@ -0,0 +1,119 @@
+
+publishing {
+    publications {
+        smartBackendAuth(MavenPublication) {
```
The name should be updated here...

dehall commented on build.gradle:
```diff
@@ -0,0 +1,119 @@
+
+signing {
+  required { project.hasProperty('signing.keyId') }
+  sign publishing.publications.smartBackendAuth
```
... and needs to match here

dehall commented on src/main/java/org/mitre/hapifhir/client/IServerClient.java:
```diff
@@ -0,0 +1,25 @@
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Bundle;
+
+public interface IServerClient {
```
Could this be given a more descriptive name, or at least more documentation on what this is or why someone would want to implement it? The name "ServerClient" isn't super obvious what it means. Seems like it's a wrapper around the built in HAPI IGenericClient to add authentication, so it would be nice if the name could reflect the auth aspect at least.

blangley28 replied:
Added a summary to the JavaDoc. Open to renaming but don't have any suggestions. Hopefully the comment makes it clear that this isn't necessarily just a wrapper around IGenericClient. Joe was researching ways to do this without making a network call. That implementation would also satisfy this interface. As we talked about before the different HAPI servers may have different ways of handling this and may want to use something other than IGenericClient

dehall commented on src/main/java/org/mitre/hapifhir/client/BearerAuthServerClient.java:
```diff
@@ -0,0 +1,43 @@
+    private String token;
+    private IGenericClient client;
+
+    public BearerAuthServerClient(String token, IGenericClient client) {
```
This is nice but it depends on a single token being valid forever, like the admin token, and things like that (hopefully) aren't used in the real world. One thing we could do to make this more generically useful is to have a second constructor take in a function to fetch a token, ie a  `java.util.function.Supplier<String>` with signature `() -> { ... return token; };`, and then the functions below would either use the single token or call the function to get one.
Or alternatively let's take this impl into the medmorph-fhir-server

blangley28 replied:
I was planning on waiting for [MEDMORPH-103](https://jira.mitre.org/browse/MEDMORPH-103) since that would implement all of those token methods. Once we have that it would be easy to re-implement this interface. I like the idea of having it pass in a Supplier but creating a new SmartBackendAuthServerClient would be nice too so others using that scheme don't need to reimplement the token methods

dehall commented on src/main/java/org/mitre/hapifhir/SubscriptionInterceptor.java:
```diff
@@ -0,0 +1,275 @@
+    @Hook(Pointcut.SERVER_OUTGOING_RESPONSE)
+    public boolean outgoingResponse(RequestDetails theRequestDetails, IBaseResource theResource) {
+        // Determine which SubscriptionTopics, if any, should be triggered
+        List<SubscriptionTopic> matchedSubscriptionTopics = getSubscriptionTopics(theRequestDetails, theResource);
```
Thoughts on filtering down to only POST/PUT/DELETE requests at the top? We know we will never send a notification on a GET so we can save some processing time by exiting early and not searching over the whole list of subscriptions. There's probably even more filtering we could do right off the bat

Commit:
PR Comments

Check request type on SubscriptionInterceptor
Add summary to IServerClient
Fix names in build.gradle

dehall committed:
copy over latest CI settings from other libraries

dehall commented on src/main/java/org/mitre/hapifhir/SubscriptionInterceptor.java:
```diff
@@ -0,0 +1,281 @@
+                try {
+                    Subscription subscription = 
+                        this.jparser.parseResource(Subscription.class, theRequestDetails.getReader());
+                    if (subscription.getStatus().equals(SubscriptionStatus.REQUESTED)) {
```
As currently written this will allow a subscription with any other status to pass right through, for example a subscription with status "active" could be passed directly. Is that intentional, or should we add some checking to ensure only status "requested" is allowed to come in?

blangley28 replied:
When creating a new subscription it should be set to requested. If we receive a subscription which isn't requested I don't know what we want to do with it so I just ignore it.
https://www.hl7.org/fhir/subscription.html#errors

dehall commented on src/main/java/org/mitre/hapifhir/SubscriptionInterceptor.java:
```diff
@@ -0,0 +1,281 @@
+     */
+    private List<Subscription> getSubscriptionsToNotify(String topicUrl, Resource theResource) {
+        myLogger.info("Checking all active subscriptions for topic " + topicUrl);
+        Bundle results = this.serverClient.searchOnCriteria("Subscription?status=active");
```
One quick thought on performance, since this function is called in a loop in `outgoingResponse` above, can we extract this call out of this function, and have the list of subscriptions be passed in as a function param?

Commit:
Extract search for all subscriptions to be outside of for loop

Commit:
Merge branch 'r5-backport' of github.com:mcode/r5-subscription-backport into r5-backport
